viomem/sys/Device.c

```diff
@@ -0,0 +1,530 @@
+
+    return status;
+}
+
+NTSTATUS 
+ViomemEvtDeviceD0Exit(
+    IN  WDFDEVICE Device,
+    IN  WDF_POWER_DEVICE_STATE TargetState
+    )
+{
+    PDEVICE_CONTEXT devCtx = GetDeviceContext(Device);
+
+    UNREFERENCED_PARAMETER(TargetState);
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Entry\n", __FUNCTION__);
+
+    PAGED_CODE();
+
+    ViomemTerminate(Device);
+	
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Return\n", __FUNCTION__);
+
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS 
+ViomemEvtDeviceD0ExitPreInterruptsDisabled(
+	IN  WDFDEVICE Device,
+    IN  WDF_POWER_DEVICE_STATE TargetState
+    )
+{
+    PDEVICE_CONTEXT       devCtx = GetDeviceContext(Device);
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Entry\n", __FUNCTION__);
+
+    PAGED_CODE();
+
+    ViomemCloseWorkerThread(Device);
+    if (TargetState == WdfPowerDeviceD3Final)
+    {
+		// Nothing here ?
+	}
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Return\n", __FUNCTION__);
+
+    return STATUS_SUCCESS;
+}
+
+BOOLEAN 
+ViomemISR(
+    IN WDFINTERRUPT WdfInterrupt,
+    IN ULONG        MessageID
+    )
+{
+    PDEVICE_CONTEXT     devCtx = NULL;
+    WDFDEVICE           Device;
+    UNREFERENCED_PARAMETER( MessageID );
+
+    Device = WdfInterruptGetDevice(WdfInterrupt);
+    devCtx = GetDeviceContext(Device);
+
+    if (VirtIOWdfGetISRStatus(&devCtx->VDevice) > 0)
+    {
+        TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INTERRUPT, "%s queueing DPC\n", __FUNCTION__);
+        WdfInterruptQueueDpcForIsr( WdfInterrupt );
+        return TRUE;
+    }
+    else
+    {
+        TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INTERRUPT, "%s No ISR indicated\n", __FUNCTION__);
+    }
+    return FALSE;
+}
+
+VOID 
+ViomemDPC(
+    IN WDFINTERRUPT WdfInterrupt,
+    IN WDFOBJECT    WdfDevice
+    )
+{
+    unsigned int          len;
+    PDEVICE_CONTEXT       devCtx = GetDeviceContext(WdfDevice);
+    PVOID                 buffer;
+
+    BOOLEAN               bHostAck = FALSE;
+    UNREFERENCED_PARAMETER( WdfInterrupt );
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Entry\n", __FUNCTION__);
+
+    WdfSpinLockAcquire(devCtx->infVirtQueueLock);
+    
+	if (virtqueue_has_buf(devCtx->infVirtQueue))
+	{
+        bHostAck = TRUE;
+    }
+    WdfSpinLockRelease(devCtx->infVirtQueueLock);
+
+    if(bHostAck)
+    {
+        KeSetEvent (&devCtx->hostAcknowledge, EVENT_INCREMENT, FALSE);
+    }
+
+    if(devCtx->Thread)
+    {
+       KeSetEvent(&devCtx->WakeUpThread, EVENT_INCREMENT, FALSE);
+    }
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Return\n", __FUNCTION__);
+}
+
+NTSTATUS 
+ViomemEnableInterrupts(
+    IN WDFINTERRUPT WdfInterrupt,
+    IN WDFDEVICE    WdfDevice
+    )
+{
+    PDEVICE_CONTEXT     devCtx = NULL;
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Entry\n", __FUNCTION__);
+
+    devCtx = GetDeviceContext(WdfDevice);
+    EnableInterrupt(WdfInterrupt, devCtx);
+    ViomemISR(WdfInterrupt, 0);

```

Comment:
Calling ISR on interrupt enable is copy paste from Balloon. This is really a "bug" in virtio and driver shouldn't behave like this.
Not a must, but I suggest adding comment referencing original BZ https://bugzilla.redhat.com/show_bug.cgi?id=1553633 and commit https://github.com/virtio-win/kvm-guest-drivers-windows/commit/139ef4d92f88a4f8370b273150a272187ea393c3